### ISSUE TITLE
FIPS compliancy

### ISSUE THREAD

**Issue by elmarco** [stefanberger/libtpms#51]
There are some concerns regarding implementation of crypto functions in libtpms in order to get FIPS compliancy.

Most of the crypto functions in libtpms come from the specification
https://trustedcomputinggroup.org/specifications-public-review/

and use openssl.

However, some crypto algorithm are open-coded, or use outdated openssl functions where better alternatives exist.

We need to identify those functions and provide alternatives when possible.
Since the code comes from the specification, we will have to discuss how the code changes can be integrated to the specification, before or after they are changed in libtpms etc.

**Comment by elmarco** [comment]
From IRC discussion with Stefan:

10:21 <stefan_berger> some of the crypt algorithms  (like key generation) were probably opened up so they can be cancelled, which is a features stemming from the hw interface and is                       available via sysfs
10:22 <stefan_berger> though I don't support cancellation in swtpm (but only in the CUSE implementation)

**Comment by stefanberger** [comment]
To avoid using those opened-up crypto algorithms that keep on reading the cancel flag we could introduce a compile time flag like `TPM_NO_COMMAND_CANCEL` and use functions that rely on calling into OpenSSL functions. I suppose this would address at least some of your concerns.

**Comment by stefanberger** [comment]
Unfortunately it's not that simple with keys that need to be regenerateable from a template. They are derived from a random number generator based on a seed an a key derivation function (KDF). For those the call into OpenSSL wouldn't work. For keys that can be truly random a call into OpenSSL would work. `CryptCreateObject()` is the function that takes the rng as parameter.

**Comment by simo5** [comment]
@stefanberger hi, I asked @elmarco to look into some of this.
Our concerns are indeed open implemenation of cryptographic algorithms like RSA, HMAC, ECDSA, etc.. we'd like to be able to replace all of them with the equivalent validated cryptographic functions provided by OpenSSL. Key generation and KDFs themselves also falls within the purview of FIPS so would need to use OpenSSL provided functionality.
What KDF do you use that you see problems with ?

**Comment by stefanberger** [comment]
Keys themselves can also have seeds and these keys can be migrated from one TPM to another and always have to generate the same child keys using their seeds. So the function testing for prime numbers has to behave exactly the same way (on different vendor's implementations) so that the child keys can be repeatably generated. So OpenSSL RSA key generation function could be used as long as it produces always the same keys as the existing algorithm does.

**Comment by stefanberger** [comment]
@simo5 [KDFa](https://github.com/stefanberger/libtpms/blob/master/src/tpm2/crypto/openssl/CryptHash.c#L739) seems to be involved with symmetric keys and[ KDFe](https://github.com/stefanberger/libtpms/blob/master/src/tpm2/crypto/openssl/CryptHash.c#L825) with ECC keys.

**Comment by simo5** [comment]
Looks like these KDFs have been implemnted in accord with the recommendations put forth in
NIST Special Publication 800-108, so they should be fine.
OpenSSL carries a number of KDFs already, and none matches thse we can propose new additions upstream.

I assume your previous comment about migration means these KDFs are set in stone and not changeable?

**Comment by simo5** [comment]
In fact, at least CryptKDFe looks like is an implementation of The Concatenation KDF decribed in SP 800 56 (I implemented this KDF in python cryptography a while ago and sounded familiar :-)

**Comment by stefanberger** [comment]
I think that's what key creation depends on and it needs to be repeatable for the same seeds using a KDF. The KDF will have to be the same then to get the same keys. From what I am hearing RSA keys cannot be derived from migrated keys.

**Comment by simo5** [comment]
Yes KDFs are deterministic, it's their purpose to be.
What do you mean with "RSA keys cannot be derived from migrated keys" ?

**Comment by stefanberger** [comment]
"NOTE 3	TPM2_CreateLoaded() can be used for creation of asymmetric keys but it may not be used for derivation of certain types of asymmetric keys. This limitation is because of the variability in algorithms for some asymmetric key types (such as RSA). "   My guess it may be related to the prime number generation or testing that the could be slightly different and lead to different RSA keys.

**Comment by stefanberger** [comment]
Though we may have the constraint that if a user creates a derived RSA key with the current implementation and then upgrades the TPM 2 implementation to the FIPS one that derived RSA key would still have to be the same as with the old implementation.

**Comment by simo5** [comment]
From the FIPS point of view, this is not a problem, as you cannot "upgrade" to FIPS.
Keys that are not created in FIPS mode cannot be used anyway, so inability to migrate keys between FIPS and non-FIPS mode would be "ok".

**Comment by stefanberger** [comment]
I looked at the HMAC fcuntionality and how it could be converted. I think the trouble lies in a function like `TPM2_HMAC_Start()` that would create an OpenSSL HMAC_CTX. Unfortunately this context is opaque so that it would prevent vTPM state suspend and resume because we wouldn't be able to take the HMAC state (it has [pointers](https://github.com/openssl/openssl/blob/6f888e05e5fa4f2f98e8f0c89f3b2cb8f9b3262d/crypto/hmac/hmac_lcl.h#L16)!) and resume it later on to finish the HMAC with `TPM2_SequenceComplete`.

**Comment by simo5** [comment]
Yes, all the new OpenSSL API is opaque and uses pointers. And there are no functions to export state.
How critical is this?

**Comment by stefanberger** [comment]
This depends on the use case of your environment. If you want to upgrade your host without having to shut down all VMs then you may either want to suspend the VMs or migrate them to another host. So in this case it would matter. It would of course be possible to maintain internal counters about how many OpenSSL contexts are open and prevent suspend only while there are some open, which may again enable suspend/resume.

**Comment by simo5** [comment]
Understood.
We may start with a lock for open HMAC sessions, and then see if upstream openssl is open to add an HMAC state export function. But perhaps simply delaying suspend until the HMAC is done is sufficient. It is unlikely that large/lengthy HMAC operations are being done via TPM anyway.

**Comment by stefanberger** [comment]
... gives a lot of control to users to prevent migration though ...

**Comment by simo5** [comment]
yeah, we'll definitely have to deal with it.

**Comment by stefanberger** [comment]
Also CMAC would need functions for exporting and importing of context.

**Comment by stefanberger** [comment]
I started a branch now where we could do this work. My first step was to convert symmetric encryption/decryption using AES. https://github.com/stefanberger/libtpms/tree/pure_openssl

**Comment by simo5** [comment]
Absolutely fantastic, looks like a clean job so far.

**Comment by stefanberger** [comment]
AES and TDES should be done.

**Comment by stefanberger** [comment]
I hope that FIPs allows setting the RNG of OpenSSL because this is the only way we could get deterministic RSA keys derived from a KDF via an RNG. I converted all the other operations of RSA to use OpenSSL functions, so being allowed to do this is critical.

**Comment by simo5** [comment]
It's probably a grey area at this point, but I will look into it once I am back from Red Hat Summit.

**Comment by stefanberger** [comment]
I have now prototyped the conversion of AES, TDES, HMAC, CMAC, and RSA functionality to use OpenSSL APIs. Left to do is EC related code. The hashing code already uses the OpenSSL library. Here's the 'fallout' from this prototyping so far:

- AES: no consequences
- TDES: no consequences
- HMAC: missing APIs in OpenSSL to support suspend/resume. Incompatibility with state that's being stored by existing code (breaks upgrade path)
- CMAC: Same as HMAC
- RSA: I haven't tested this yet, but the deterministically created keys in the existing code could very well be different than those created by OpenSSL (breaks upgrade path).

So, my biggest worry here is the upgrade path. One solution may be to use different compile time options for different distros... :-(

**Comment by simo5** [comment]
So the real breaker for upgrade seems to be mostly the RSA key derivation.
HMAC and CMAC are just suspend/resume issues as far as I can see, but not necessarily upgrade issues (unless you upgrade while the machine is suspended).

**Comment by stefanberger** [comment]
Yes, RSA key derivation breaks it.

'Operation in flight' is the occasion when an HMAC and CMAC sequences may have started before a VM suspend and would be expected to be completed after the resume, though because of the upgrade they either cannot resume or return the wrong result.

**Comment by stefanberger** [comment]
One possibility could be to introduce some flag that indicates whether a seed was created under the new code and if so uses the new key creation method, otherwise uses the old one... not nice for the code, though.

**Comment by simo5** [comment]
So I am looking at the code, and I am not sure there is value in trying to use OpenSSL's function to derive RSA keys.
Using a non CSPRNG random number generator is not going to be FIPS compliant anyway.

**Comment by simo5** [comment]
Let me elaborate more, what you are doing is not really generating RSA keys, but deriving keys from some other state. This is an operation that is different enough from a generation function to warrant its own function anyway. So given that change would be backwards incompatibile anyway, I would dealy implementing it until we understand what the implications of doing it under FIPS are going to be.

**Comment by stefanberger** [comment]
There are truly random numbers if rand parameter is NULL. In this case the existing code uses OpenSSL's function. We could switch this one over to use OpenSSL's function to generate a key, if this helps, leave the other case untouched.

**Comment by simo5** [comment]
Indeed I was going to ask in what case you would ever want to derive RSA or other asymmetric keys, are there any docs around this topic ?

**Comment by stefanberger** [comment]
TPM2_CreatePrimary always derives keys from seeds until the seed(s) are changed:

https://github.com/stefanberger/libtpms/blob/master/src/tpm2/HierarchyCommands.c#L102

The `CryptCreateObject()` called later on creates the key.

Here are these seeds: https://github.com/stefanberger/libtpms/blob/master/src/tpm2/Global.h#L618

Here's a doc for the TPM 2 commands. See section on TPM2_CreatePrimary: https://trustedcomputinggroup.org/wp-content/uploads/TPM_Rev2.0_Part3_Commands_r1.50-code_publicreview.pdf

**Comment by simo5** [comment]
Thanks for the pointers, but I was more looking for what are the reasons for using a seed, rather then simply just creating and storing actual RSA keys ?

**Comment by stefanberger** [comment]
One reason is that the hardware TPM doesn't have much memory to persist the private keys into, so they chose this route to always be able to recreate them rather than store them. And the primary keys are the root keys of hierarchies.

**Comment by simo5** [comment]
@stefanberger yes make sense for hardware based TPMs, are you maintaining this scheme in the software library for compatibility? Are there uses cases where you would write seeds generated in the software library back into an HW TPM ?

**Comment by stefanberger** [comment]
Yes, for full compatibility with the hardware TPM commands and usage model I am maintaining the scheme also in the software library since there are commands that can be used to change the seeds, which makes whole hierarchies of keys useless, of course. The seeds cannot be written into a hardware TPM since they are kept inside the TPM, thus are invisible to the user.

**Comment by simo5** [comment]
Just as an hypotethical, wouldn't it be sufficient to simply drop all stored keys when the seed is changed to maintain command/behavior compatibility ?

**Comment by stefanberger** [comment]
Sure. Some keys live outside of the TPM as child keys to the primary keys. These child keys then won't load anymore and a user has to clean them up from the filesystem.

**Comment by stefanberger** [comment]
... there are also persisted keys inside the TPM 2. I believe they are not cleaned up automatically when a seed changes but the user will have to delete them upon detection that they are now useless.

**Comment by stefanberger** [comment]
Also EC keys created with OpenSSL when derived from a KDF turn out to be different than those created with the TPM 2 algorithms.

**Comment by simo5** [comment]
I see you replace the random function in OpenSSL, does that create different keys even if you use the same random function currently used by TPM 2?

**Comment by stefanberger** [comment]
It creates different keys. I'll do some digging into the EC private keys generated this way and how the byte 'patterns' of the private keys differ between the two methods.

**Comment by stefanberger** [comment]
Using OpenSSL it asks for 32 bytes 3 times and takes the 1st 32 bytes as the key:
```
Getting 32 bytes
25 e0 3e 44 ed 2b bd ce ea bc 98 87 8b ed 18 d5 ab 0f 25 24 18 0e 81 14 43 a0 a3 9c 9d fe 65 27 
Getting 32 bytes
42 2d 6c 96 4c 32 d8 99 4c 77 84 e7 88 62 ad 21 24 00 6f c6 b2 49 00 68 7c 9e 75 8d 57 73 8a 49 
EC private key D: 25e03e44ed2bbdceeabc98878bed18d5ab0f2524180e811443a0a39c9dfe6527
Getting 32 bytes
e2 94 c7 3c 5d aa 96 d7 ab 9d de d5 51 75 4e c0 4b a5 73 50 08 a6 cd 7c c0 4d d7 d1 78 1b e4 a0 
```

Using the TPM2 code it asks for 40 random bytes and then does some additional math with the random number: https://github.com/stefanberger/libtpms/blob/master/src/tpm2/crypto/openssl/CryptEccMain.c#L561
```
Getting 40 random bytes
25 e0 3e 44 ed 2b bd ce ea bc 98 87 8b ed 18 d5 ab 0f 25 24 18 0e 81 14 43 a0 a3 9c 9d fe 65 27 8c c5 e8 fd b3 c5 db f6 
(old) EC private key D: d7e8565578e11cc2b4fc891fd0fc42c4898b1fba001f73832385497bd53f5707
```

**Comment by simo5** [comment]
Ok, then it sounds like the short term approach may be to have a "fips mode" and conditionalize what is used based on that, I will try to investigate if there is a way to use a DRBG that matches what TPM2 is using.

**Comment by stefanberger** [comment]
@simo5 Does FIPs require freed memory to be cleared? What about possibly sensitive variables on the stack?

**Comment by simo5** [comment]
@stefanberger yes FIPS generally indicates that memory holding any type of key should be zeroed as son as it is not used anymore, for example immediately before a free(), or when the memory should be returned to the garbage collector for languages that use that kind of memory management.

**Comment by stefanberger** [comment]
@simo5 I am not sure how to best go about this with on-stack variables after reading this conversation here where they say `memset()` is being optimized out if it's put at the end of a function: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=69976   Ideally gcc would have something like this and better than the `__attribute__ ((__cleanup__ ...))` function now. It looks like it would have to be a function called by a different name than `memset()`?

**Comment by simo5** [comment]
@stefanberger 
Recent glibc provides the explicit_bzero() function that is guaranteed never to be optimized out.
The C11 standard defines memset_s() and explicitly requires it cannot be optimized out, it is unclear to me how available this function is yet on various compilers.
I forget if you need to be portable in that case Microsoft offers SecureZeroMemory() on Windows.

**Comment by stefanberger** [comment]
@simo5 @elmarco can you have a look at PR https://github.com/stefanberger/libtpms/pull/58. I am just clearing all EC_POINT and BIGNUMs that so far are not cleared. I'd appreciate help in finding the places that need more clearing, including variables on the stack that need clearing. For this at least I now started adding a CLEARMEM #define in this series now: https://github.com/stefanberger/libtpms/commits/pure_openssl .

**Comment by elmarco** [comment]
@stefanberger looks good to me. I am a bit surprised that openssl doesn't seem to care about clearing variables. For clearing the stack, there are compiler tricks https://stackoverflow.com/questions/29320992/gcc-linux-set-the-stack-to-zero-before-returning, but that's compiler-specific.

In libtpms, there is MemorySet() called from sensible places. You may want to replace it with your CLEARMEM?

**Comment by simo5** [comment]
@stefanberger I think the patch is ok, the discussion about cleraing the stack is a bit complicated. Stack variable tend to be overwritten pretty quickly in an active program. Although it is true that a call that ends up to be deeply nested can be pushed so high in the stack that data lingers there for a significant amount of time, that is not considered as bad as leaving unzeroed data in the heap where they can definitely linger for a long time and also be accidentally leaked more easily.

That said if there is an easy way to zero stack variables, like in #58 case, it is a good idea to do so.

**Comment by stefanberger** [comment]
@elmarco @simo5 Thanks for looking at it. I'll need you guys for reviewing...

I dropped the CLEARMEM now in that other series and won't go crazy about sensitive data on the stack...

For the code conversion to use OpenSSL functions, I would like to first only do those conversions where there are no side effects to upgrade path and state. I am not sure how to go about the other parts then, though. Whether there will be a configure option for FIPS compliance via OpenSSL function calls. The problems are obviously with key creations from KDFs.

**Comment by simo5** [comment]
Sorry, I haven't had time yet to carefully look into the KDFs, however your sounds like a good plan.

**Comment by stefanberger** [comment]
So here's the PR for the symmetric key operations with AES and TDES keys and the various modes: https://github.com/stefanberger/libtpms/pull/61

**Comment by stefanberger** [comment]
And another small PR: https://github.com/stefanberger/libtpms/pull/64

**Comment by stefanberger** [comment]
Aeh, this one for ECDSA part 1: https://github.com/stefanberger/libtpms/pull/63

**Comment by simo5** [comment]
Will look at #63 shortyl, meanwhile can you rename this issue as FIPS Compliance, it's been bothering me for a while that we use a non-existing word :-)

**Comment by stefanberger** [comment]
RSA encryption and signing code is now in this PR: https://github.com/stefanberger/libtpms/pull/65

**Comment by stefanberger** [comment]
Any comments on PR #65 would be appreciated...

**Comment by stefanberger** [comment]
Non-KDF EC key generation is now here: https://github.com/stefanberger/libtpms/pull/67

**Comment by stefanberger** [comment]
There are 3 seeds (EPSeed, SPSeed, and PPSeed) in the TPM 2.0 that are used for feeding the KDFs which in turn are used for deterministically creating keys. I did some prototyping around tracking these seeds for when they are created. We need a flag for each seed, which we set in case the seed is newly created, which then indicates that an OpenSSL function can be used for deriving this key, while if the flag is not set we have to use the original TPM 2 code for creating the key. Today we only have the latter case. The seed creation tracking is needed for the upgrade path.

The prototype is in this branch here: https://github.com/stefanberger/libtpms/tree/pure_openssl_seeds

Any key that is a child key of a primary key would have to inherit the `newSeed` flag (see code) from its parent.

**Comment by stefanberger** [comment]
One of the issues is whether we can create all necessary key types in OpenSSL while deriving them from a KDF. Playing around with OpenSSL's RNG for this purpose may not be allowed...

**Comment by simo5** [comment]
It is allowed to use a DRBG in this manner, but we need to use the correct CSPRNG.

CCing Tomáš, he knows the code better than I around OpenSSL RNGs

@t8m can you help ?

**Comment by stefanberger** [comment]
Small patch for TDES key creation when no KDF is used: https://github.com/stefanberger/libtpms/pull/68

**Comment by t8m** [comment]
What is the requirement for the KDF? Why not use some preexisting KDF to derive the keys? I.e. do not use a fixed seed to DRBG but instead use a regular KDF such as HKDF.

**Comment by simo5** [comment]
t8m, keys are derived from the seed each time the "device" is "restarted", the intention was to find out if we can be compatible with previous code, so that existing seeds keep "working" (ie generate the same keys as before).

**Comment by t8m** [comment]
But what was the previous code doing to derive the keys? Did it somehow depend on the exact workings of OpenSSL rng? Or did it use its own DRBG implementation to derive the keys? Or?

**Comment by stefanberger** [comment]
The original TPM 2 code use an RNG backed by a KDF (rather than /dev/urandom) to generate the keys. The original TPM 2 code creates primes itself, tested them, etc. so that no OpenSSL functions other than BIGNUM related ones were used.

**Comment by t8m** [comment]
I am afraid that here we are out of ways that could be possibly FIPS compliant. The only current FIPS approved method for generating keys based on a preselected seed that we implement in OpenSSL is the DSA FIPS 186-3 based keygen. But that does not help you. We could ask a FIPS test lab what is their opinion about generating RSA/ECDSA keys by some similar method. But then anyway it would have to be implemented in the OpenSSL directly and not outside of it (otherwise the libtpms would have to be inside a FIPS module boundary).

**Comment by simo5** [comment]
@t8m baby steps, let's get as close as we can, then we'll make a summary of what may still not be compliant and we'll see what can be done either by more breaking changes enabled only in FIPS mode or by opening a conversation with NIST via our Lab on these key derivation issues.
I have to assume the key derivation itself is something that would be interesting to NIST because a lot of actual TPMs must work that way in order to conserve memory.

**Comment by t8m** [comment]
Is it possible they do not use key derivation but key wrapping? I.e. the RSA/ECDSA key is generated randomly but wrapped with a key stored in TPM.

**Comment by stefanberger** [comment]
No, they use key derivation in many cases, such as for the root keys of hierarchies (TPM2_CreatePrimary) as well as other keys that are child keys (TPM2_CreateLoaded).

https://trustedcomputinggroup.org/wp-content/uploads/TCG_TPM2_r1p55_Part3_Commands_code_pubrev.pdf

**Comment by stefanberger** [comment]
> But then anyway it would have to be implemented in the OpenSSL directly and not outside of it (otherwise the libtpms would have to be inside a FIPS module boundary).

This sounds like OpenSSL's crypto library will have to support the KDFs that TPM 2 defines as well as provide APIs to derive any type of key (RSA, EC, AES, Camellia, TDES) from those KDFs, since this is what TPM 2 also supports. I hope that the OpenSSL community is receptive to such an extension...

**Comment by simo5** [comment]
@stefanberger I think we should worry about perfect compliance later, and for now get as closer as we can get to using a crypto module (OpenSSL) so that we greately reduce the "gap" between what is compliant and waht is not.
As for KDFs and RNGs NIST has a specific set of allowed algorithms, so we'll have to look into it more carefully once we figure what can be used from OpenSSL, what can, and what creates issues (eg, HMAC and saving state, or the KDF).
There is always an option that companies may decide to certify libtmps or a vTPM componet as a cryptographic module of its own, in that case some cryptography (like the KDF) can be implemented in the library, but of course still offloaing as much as possible to OPenSSL will greately reduce the scope of work (and cost).

So for now I would simply note the issues, and work on low hanging fruits and make a summary once we know all problematic corners so that a plan can be developed on how to best address those.

**Comment by stefanberger** [comment]
@simo5 I think the pending PRs get us as to the first major hurlde of what can be done with libcrypto without having to sacrifice functionality in libtpms, such as for example state suspend/resume support, which is the reason why I am not including HMAC/CMAC in such PRs.

**Comment by stefanberger** [comment]
I revised the RSA PR a bit: https://github.com/stefanberger/libtpms/pull/65 Unless there are comments I would merge by the end of the week. There are some performance losses due to having to recreate the private exponent D all the time, but this could be addressed separately if performance was a concern.

**Comment by simo5** [comment]
Sooo rsaKey does not include D and needs to be recomputed every time ?

**Comment by stefanberger** [comment]
We haven't stored D so far because the original TPM 2 algorithms didn't need it and used the Chinese Remainder Theorem (CRT) parameters for private keys and its operations, which seems to speed things up. OpenSSL unfortunately also needs the D parameter, which could be a bug, but it doesn't help since we would crash if we don't provide it...

**Comment by stefanberger** [comment]
PR https://github.com/stefanberger/libtpms/pull/71 now implements support for the two opposite key operations: decryption and signature verification

**Comment by simo5** [comment]
> We haven't stored D so far because the original TPM 2 algorithms didn't need it and used the Chinese Remainder Theorem (CRT) parameters for private keys and its operations, which seems to speed things up. OpenSSL unfortunately also needs the D parameter, which could be a bug, but it doesn't help since we would crash if we don't provide it...

Uhmm indeed it could be argued that if the CRT parameters are provided then d shouldn't be necessary, openssl uses the crt as well iirc ... anyway I understand why you need to add it now which is what I needed.

**Comment by stefanberger** [comment]
PR https://github.com/stefanberger/libtpms/pull/72 now uses OpenSSL for creating purely random keys (if rand == NULL) just like it does for TDES and EC keys.

**Comment by stefanberger** [comment]
I believe the merged PRs reflect now all the changes that can be done with OpenSSL at the moment. So the biggest issue now is:
- key derivation cannot be cleanly done with OpenSSL for the key types (RSA, EC, AES, TDES, etc.) and (custom) KDFs supported by TPM 2

**Comment by simo5** [comment]
@stefanberger yes I think this was a very nice piece of work, thanks a lot for it.
For the next steps I am planning to talk with our certification lab when I have time to see what are viable options wrt KDF, saving state, etc...

**Comment by stefanberger** [comment]
Ok.

FYI: The next patch series is an upgrade to revision 155  of the TPM 2 specs (PR https://github.com/stefanberger/libtapms/pull/70). I may have to delay this to rev 156 or so, though. There were some issues with RSA key generation on 64 bit machines, which is fixed there. To take this code and to maintain backwards compatibility I have to start tracking when seeds are generated and use the new code for new seeds while dragging along the old key generation code to keep on generating the same keys... The same seed tracking needs to then be done for switching to OpenSSL key generation, if this was to happen.

**Comment by stefanberger** [comment]
@simo5 Is there a requirement on what OpenSSL crypto APIs an application is supposed to or must use to (easier) become FIPs compliant? Like one should/must only use `EVP_*` functions?

**Comment by simo5** [comment]
In general EVP is strongly preferred where possible.

**Comment by nmav** [comment]
@stefanberger I see that what you are describing in terms of RSA key generation, is very close to the Shawe-Taylor algorithm. That's a FIPS approved algorithm (FIPS186-4), and generates RSA keys from a seed (and is additionally a provable key generation method). gnutls uses this algorithm internally when provable keys are requested (e.g, `certtool --generate-privkey --provable --seed "123456"` will generate always the same key). I do not know whether openssl supports this algorithm though.

**Comment by t8m** [comment]
Unfortunately openssl does not implement this algorithm.

**Comment by elmarco** [comment]
Is there a todo-list of things known left to work on? (sorry I didn't read the whole thread)

**Comment by stefanberger** [comment]
Afaik I did everything that can be done for changing TPM 2 code to use OpenSSL functions.

**Comment by stefanberger** [comment]
> Is there a todo-list of things known left to work on? (sorry I didn't read the whole thread)

TCG published a FIPS 140-3 guidance document: https://trustedcomputinggroup.org/resource/tcg-fips-140-3-guidance-for-tpm-2-0/

There's an old document here: https://trustedcomputinggroup.org/resource/tcg-fips-140-2-guidance-for-tpm-2-0/

Known issues:
- KDFa is a bigger issue the way it is used by the TPM 2 reference code. It's long-term stateful and can be resumed at any later point in time but the OpenSSL KBKDF cannot be resumed. Prototype for it using OpenSSL KDF function is in this PR: https://github.com/stefanberger/libtpms/pull/426  Basically it cannot use OpenSSL's KBKDF
- KDFe could be using OpenSSL SSKDF: https://github.com/stefanberger/libtpms/pull/424
- Pairwise consistency tests are only implemented for EC signing keys
- ...
